File: bdd/cpp/features/step_definitions/raw_command_steps.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#define CUKE_OBJECT_PREFIX IggyBddRawCommand

#include <gtest/gtest.h>

#include <cucumber-cpp/autodetect.hpp>

#include <cstdint>
#include <exception>
#include <string>

#include "world.hpp"

WHEN("^I send a raw command with code ([0-9]+) and an empty payload$") {
    REGEX_PARAM(std::uint32_t, code);
    cucumber::ScenarioScope<bdd::GlobalContext> context;

    context->raw_response.clear();
    context->raw_error.clear();
    try {
        const auto response = context->client->send_binary_request(code, rust::Vec<std::uint8_t>());
        context->raw_response.assign(response.begin(), response.end());
    } catch (const std::exception &error) {
        context->raw_error = error.what();
    }
}

THEN("^the raw command should succeed with an empty response$") {
    cucumber::ScenarioScope<bdd::GlobalContext> context;

    EXPECT_TRUE(context->raw_error.empty());
    EXPECT_TRUE(context->raw_response.empty());
}

THEN("^the raw command should succeed with a non-empty response$") {
    cucumber::ScenarioScope<bdd::GlobalContext> context;

    EXPECT_TRUE(context->raw_error.empty());
    EXPECT_FALSE(context->raw_response.empty());
}

THEN("^the raw command should fail with an invalid command error$") {
    cucumber::ScenarioScope<bdd::GlobalContext> context;

    EXPECT_TRUE(context->raw_response.empty());
    EXPECT_NE(context->raw_error.find("Invalid command"), std::string::npos);
}

File: foreign/csharp/Iggy_SDK.Tests.BDD/StepDefinitions/RawCommandSteps.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using Apache.Iggy.Exceptions;
using Reqnroll;
using Shouldly;
using TestContext = Apache.Iggy.Tests.BDD.Context.TestContext;

namespace Apache.Iggy.Tests.BDD.StepDefinitions;

[Binding]
public class RawCommandSteps
{
    private readonly TestContext _context;

    public RawCommandSteps(TestContext context)
    {
        _context = context;
    }

    [When(@"I send a raw command with code (\d+) and an empty payload")]
    public async Task WhenISendARawCommand(uint code)
    {
        try
        {
            _context.LastRawResponse = await _context.IggyClient.SendBinaryRequestAsync(code, []);
            _context.LastRawError = null;
        }
        catch (Exception error)
        {
            _context.LastRawResponse = null;
            _context.LastRawError = error;
        }
    }

    [Then(@"the raw command should succeed with an empty response")]
    public void ThenTheRawCommandShouldSucceedWithAnEmptyResponse()
    {
        _context.LastRawError.ShouldBeNull();
        _context.LastRawResponse.ShouldBeEmpty();
    }

    [Then(@"the raw command should succeed with a non-empty response")]
    public void ThenTheRawCommandShouldSucceedWithANonEmptyResponse()
    {
        _context.LastRawError.ShouldBeNull();
        _context.LastRawResponse.ShouldNotBeEmpty();
    }

    [Then(@"the raw command should fail with an invalid command error")]
    public void ThenTheRawCommandShouldFailWithAnInvalidCommandError()
    {
        _context.LastRawResponse.ShouldBeNull();
        var error = _context.LastRawError.ShouldBeOfType<IggyInvalidStatusCodeException>();
        error.StatusCode.ShouldBe(3);
    }
}
